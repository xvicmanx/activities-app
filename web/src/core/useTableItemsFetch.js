// @flow

import { useState, useEffect } from 'react';

import type { Result } from './helpers';
import { DEFAULT_RESULT } from './helpers';

const useTableItemsFetch = (fn: Function, payload: Object): Result => {
  const [result, setResult] = useState(DEFAULT_RESULT);

  useEffect(() => {
    const fetchData = async () => {
      const result = await fn(payload);
      setResult(result);
    };  
    fetchData();
  }, [fn, payload]);

  return result;
};

export default useTableItemsFetch;
