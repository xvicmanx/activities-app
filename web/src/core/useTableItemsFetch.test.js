import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import useTableItemsFetch from './useTableItemsFetch';

Enzyme.configure({ adapter: new Adapter() });

// eslint-disable-next-line react/prop-types
const TestCmp = ({ cb }) => {
  const result = useTableItemsFetch(cb);
  return <div>{JSON.stringify(result)}</div>;
};

describe('useTableItemsFetch', () => {
  it('works as expected', () => {
    const items = [
      {
        id: 1,
        name: 'test',
      },
    ];
    const promiseResult = { items, total: items.total };
    const cb = jest.fn(() => Promise.resolve(promiseResult));

    jest.spyOn(React, 'useEffect');
    jest.spyOn(React, 'useState');

    const component = Enzyme.mount(<TestCmp cb={cb} />);

    expect(React.useEffect).toHaveBeenCalledTimes(1);
    expect(React.useState).toHaveBeenCalledTimes(1);

    expect(cb).toHaveBeenCalledTimes(1);
    expect(component.html()).toMatchSnapshot();
  });
});
