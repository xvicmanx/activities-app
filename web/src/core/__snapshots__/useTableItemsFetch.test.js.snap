// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTableItemsFetch works as expected 1`] = `"<div>{\\"total\\":0,\\"items\\":[]}</div>"`;
